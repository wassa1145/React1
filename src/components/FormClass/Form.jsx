import React, { Component } from 'react';
import { Button } from './Button';
import { Textarea } from './Textarea';
import { Message } from './Message';

export class Form extends Component {
	state = {
		name: 'Отправить',
		value: '',
		message: 'Привет! Я дефолтное сообщение. Можешь меня заменить',
	}
	handleClick = () => {
		this.setState({ message: this.state.value })
		this.setState({ value: '' })
	}
	handleChange = (event) => {
		this.setState({ value: event.target.value })
	}
  render() {
		return <div className='form'>
			<Message message={this.state.message}/>
			<Textarea change={this.handleChange} value={this.state.value}/>
			<Button name={this.state.name} click={this.handleClick} disabled={this.state.value ? false : true}/>
		</div> 
  }
}
