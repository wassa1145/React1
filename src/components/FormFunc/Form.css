.form {
  width: 100%;
  padding: 20px;
  margin: 0 auto;
  box-sizing: border-box;
	margin-top: 30px;
  display: flex;
  justify-content: space-between;
}
