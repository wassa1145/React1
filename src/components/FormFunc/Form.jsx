import React, { useState, useEffect, useRef } from 'react';
import { Button } from './Button';
import { Textarea } from './Textarea';
import { Input } from './Input';
import { List } from './List';

export const Form = () => {
  const [name] = useState('Отправить');
  const [inputPlaceholder] = useState('Введите ваше имя');
  const [defaultMessage] = useState(
    'Мы получили ваше сообщение! В ближайшее время мы вам ответим.'
  );
  const [defaultAuthor] = useState('Mr. Robot');
  const [textareaPlaceholder] = useState('Введите сообщение');
  const [message, setMessage] = useState('');
  const [author, setAuthor] = useState('');
  const [messageList, setMessageList] = useState([]);
  const listEl = useRef(null);

  const handleClick = () => {
    setMessageList([...messageList, { message, author, systemMessage: false }]);
    setMessage('');
    setAuthor('');
  };
  const changeАuthor = (event) => {
    setAuthor(event.target.value);
  };
  const changeMessage = (event) => {
    setMessage(event.target.value);
  };
  const scrollList = () => {
    listEl.current.scrollTop = listEl.current.scrollHeight;
  };
  useEffect(() => {
    if (
      messageList.length &&
      !messageList[messageList.length - 1].systemMessage
    ) {
      setTimeout(() => {
        setMessageList([
          ...messageList,
          {
            message: defaultMessage,
            author: defaultAuthor,
            systemMessage: true,
          },
        ]);
      }, 1000);
    }
    scrollList();
  }, [messageList.length]);

  return (
    <div className="content">
      {<List messages={messageList} refEl={listEl} />}
      <div className="form">
        <div className="form__wrapper">
          <Input
            placeholder={inputPlaceholder}
            author={author}
            change={changeАuthor}
          />
          <Textarea
            change={changeMessage}
            message={message}
            placeholder={textareaPlaceholder}
          />
        </div>
        <Button
          name={name}
          click={handleClick}
          disabled={message && author ? false : true}
        />
      </div>
    </div>
  );
};
