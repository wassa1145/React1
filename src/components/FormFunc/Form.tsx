import React, { useState, useEffect, FC, memo, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { ThunkDispatch } from 'redux-thunk';
import { ChatsState } from 'src/store/chats/reducer';
import { AddMessage } from 'src/store/chats/types';
import { CONSTANTS } from 'src/constants';
import { addMessageWithReply } from 'src/store/chats/slice';
import { Button } from 'src/components/FormFunc/components/Button/Button';
import { Textarea } from 'src/components/FormFunc/components/Textarea/Textarea';
import './Form.css';

export const Form: FC = memo(() => {
  const [message, setMessage] = useState('');
  const textFieldRef = useRef<HTMLInputElement>(null);
  const { chatId } = useParams();
  const dispatch =
    useDispatch<ThunkDispatch<ChatsState, void, ReturnType<AddMessage>>>();

  const handleClick = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (chatId && message) {
      dispatch(
        addMessageWithReply({
          chatId,
          message: { author: CONSTANTS.USER, value: message },
        })
      );
    }
    setMessage('');
    textFieldRef.current?.focus();
  };

  useEffect(() => {
    textFieldRef.current?.focus();
  }, []);

  return (
    <form className="form" onSubmit={handleClick} data-testid="form">
      <Textarea
        change={(event) => setMessage(event.target.value)}
        message={message}
        textFieldRef={textFieldRef}
        placeholder="Введите сообщение"
      />
      <Button name="Отправить" disabled={!message} />
    </form>
  );
});
