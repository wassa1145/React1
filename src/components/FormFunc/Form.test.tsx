import React from 'react';
import { Form } from './Form';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

beforeEach(() => {
  render(<Form />);
});

describe('Form', () => {
  it('render component', () => {
    expect(screen.getByTestId('form')).toBeInTheDocument();
  });

  it('render with snapshot', () => {
    const { asFragment } = render(<Form />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('render textarea and button', () => {
    expect(screen.getByTestId('textarea')).toBeInTheDocument();
    expect(screen.getByTestId('submit-button')).toBeInTheDocument();
  });

  it('clear textarea on subbmit', async () => {
    await userEvent.type(screen.getByRole('textbox'), 'new message');
    await userEvent.click(screen.getByTestId('submit-button'));
    expect(screen.getByRole('textbox')).toHaveValue('');
  });
});
