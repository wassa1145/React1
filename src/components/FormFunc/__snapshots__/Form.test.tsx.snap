// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render with snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
    data-testid="form"
  >
    <div
      class="MuiFormControl-root MuiTextField-root form__textarea css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="textarea"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r2:"
          placeholder="Введите сообщение"
          rows="4"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-hdw1oc"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root Mui-disabled form__button css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      data-testid="submit-button"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      Отправить
    </button>
  </form>
</DocumentFragment>
`;
